### PR TITLE
Issue 3801 from Roslyn project

### PR DESCRIPTION
**Pull request by hvanbakel**
Fixed issue 3801 in the roslyn project:
https://github.com/dotnet/roslyn/issues/3801

Moved pull request here

### CODE REVIEW

**Comment by tmeschter** [comment]
@srivatsn @mavasani @lgolding Yay! First community pull request for dotnet/roslyn-analyzers! Who wants to review it?

**Comment by hvanbakel** [comment]
@srivatsn already did it here:
https://github.com/dotnet/roslyn/pull/4322

**Comment by srivatsn** [comment]
Functionally it looks good. @hvanbakel  Are you keen on making the formatting changes? Generally stylistic preference tends to be subjective and it's best to keep things consistent. So we have a codeformatter tool that we run to fix up formatting\style that we run once in a while. So, unless you think these changes make a large difference, I'd revert them and keep this PR to be about just the functional change.

File: src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -13,59 +13,84 @@ public sealed class SerializationRulesDiagnosticAnalyzer : DiagnosticAnalyzer
...
+            isEnabledByDefault: true,
+            description: s_localizableDescriptionCA2235,
+            helpLinkUri: "http://msdn.microsoft.com/library/ms182349.aspx",
+            customTags: DiagnosticCustomTags.Microsoft);
 
-        private static readonly ImmutableArray<DiagnosticDescriptor> s_supportedDiagnostics = ImmutableArray.Create(RuleCA2229, RuleCA2235, RuleCA2237);
+        private static readonly ImmutableArray<DiagnosticDescriptor> s_supportedDiagnostics =
+            ImmutableArray.Create(RuleCA2229, RuleCA2235, RuleCA2237);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
```
Comment (srivatsn) [inline src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs]: Instead use expression bodied member: => s_supportedDiagnostics and that can be inlined as well and you can get rid of the field

**Comment by hvanbakel** [comment]
All the style changes where done by visual studio/resharper. I can revert them if necessary.

**Comment by srivatsn** [comment]
In that case yes, it would be cleaner to have this change just be about the bug fix.

File: src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -186,7 +234,34 @@ public void AnalyzeSymbol(SymbolAnalysisContext context)
...
+                       namedTypeSymbol.GetAttributes()
+                           .Any(a => a.AttributeClass.Equals(_serializableAttributeTypeSymbol));
+            }
+
+            private static bool IsPrimitiveType(ITypeSymbol type)
+            {
+                switch (type.SpecialType)
+                {
+                    case SpecialType.System_Boolean:
+                    case SpecialType.System_Byte:
```
Comment (lgolding) [inline src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs]: Missing `System_Decimal`?

Comment (hvanbakel) [inline src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs]: In that case it has to be fixed here as well:
http://source.roslyn.io/#System.Runtime.Analyzers/Reliability/DoNotLockOnObjectsWithWeakIdentity.cs,92432dae4f408c01

Comment (srivatsn) [inline src/AnalyzerPowerPack/Core/Usage/SerializationRulesDiagnosticAnalyzer.cs]: Decimal is not a primitive type (the CLR doesn't understand decimals - it's a c# keyword). So this method shouldn't include it but we probably don't want to complain about decimal fields and so the check should be next to the string check.

**Comment by hvanbakel** [comment]
All comments fixed

**Comment by srivatsn** [comment]
Thanks.

**Comment by 333fred** [comment]
@pareshjoshi could you please file a new issue with what you're seeing?

**Pull request merged**